lib/WorkflowOncoanalyser.groovy

```diff
@@ -2,76 +2,125 @@
 // This file holds several functions specific to the workflow/oncoanalyser.nf in the nf-core/oncoanalyser pipeline
 //
 
-import groovy.text.SimpleTemplateEngine
+import static groovy.io.FileType.FILES
+
+import nextflow.Channel
+import nextflow.Nextflow
+
+import Constants
+import Processes
+import Utils
 
 class WorkflowOncoanalyser {
 
-    //
-    // Check and validate parameters
-    //
-    public static void initialise(params, log) {
-        genomeExistsError(params, log)
+    public static groupByMeta(Map named_args, ... channels) {
```

Comment:
can you explain a bit more what these functions can do that you can't achieve by build-in nextflow functions?

Reply:
Sure! The purpose of these functions primarily to reduce verbosity of the code and to improve maintainability and readability e.g. you see the function name and immediately understand intention rather than deciphering a dozen or so of lines of code, which may be subtlety different from the other 40+ instances of this operation / pattern.

**groupByMeta** and co supports non-blocking and ordered joining of two or more channels - as far as I'm aware doing both of these tasks is not readily possible using the built-in NF operators and require additional operations to achieve this goal. These additional operations are implemented once within groupByMeta to provide the desired high-order functionality.

**restoreMeta** and co is a convenience function that flexibly handles a process output channel where the meta has been minimised to avoid needlessly invalidating NF cache on unrelated input/parameter change. Specifically, the restoreMeta function takes an output channel and replaces the minimised meta with the complete meta for seemless downstream processing. This is all done using built-in NF operators (except for the flexibly selecting the join key from the meta) but implemented within a function to reduce repetitiveness and improve readability.